pkg/distributor/ha_tracker.go

```diff
@@ -0,0 +1,222 @@
+			return &ReplicaDesc{Replica: replica, ReceivedAt: timestamp.FromTime(now)}, false, nil
+		}
+
+		// As far as the timestamp in the KVStore is concerned none of the Distributors have
+		// received a sample from the stored replica recently, so we should overwrite the
+		// replica in the KVStore and accept this sample.
+		if now.Sub(timestamp.Time(desc.ReceivedAt)) > c.failoverTimeout {
+			desc.ReceivedAt = timestamp.FromTime(now)
+			desc.Replica = replica
+			return desc, false, nil
+		}
+
+		// Safeguard against other calls to the KVStore having changed the accepted replica via a
+		// CAS to the KVStore before us. We should only return a replica description when the
+		// replicas don't match in the case above, where the overwrite timeout has been reached.
+		if desc.Replica != replica {
+			// Return a 202.
+			return nil, false, httpgrpc.Errorf(http.StatusAccepted, "replicas did not match, rejecting sample: %s != %s", replica, desc.Replica)
+		}
+
+		// As far as the timestamp in the KVStore is concerned none of the Distributors have
+		// updated the timestamp for this cluster/replica in more than the write timeout.
+		// We should update the timestamp so that we don't potentially failover to accepting
+		// samples from another replica when we don't need to.
+		// This may only catch local CAS' updating the KVStore, not other Distributors.
+		if now.Sub(timestamp.Time(desc.ReceivedAt)) > c.updateTimeout {
+			desc.ReceivedAt = timestamp.FromTime(now)
+			return desc, false, nil
+		}
+		return nil, false, nil
+	})
+	return err
+}
+
+// CheckReplica checks the cluster and replica against the backing KVStore and local cache in the
+// tracker c to see if we should accept the incomming sample. It will return an error if the sample
+// should not be accepted. Note that internally this function does checks against the stored values
+// and may modify the stored data, for example to failover between replicas after a certain period of time.
+func (c *tracker) CheckReplica(ctx context.Context, cluster, replica string) error {
+	cas := false
+	now := mtime.Now()
+	c.electedLock.RLock()
+	entry, ok := c.elected[cluster]
+	c.electedLock.RUnlock()
+
+	// No replica in local cache for this HA cluster yet.
+	if !ok {
+		return c.checkKVStore(ctx, cluster, replica, now)
+	}
+
+	// The replica was in the distributors cache.
+	if entry.Replica != replica {
+		if now.Sub(timestamp.Time(entry.ReceivedAt)) < c.failoverTimeout {
+			// We don't want to overwrite yet.
+			return fmt.Errorf("error bro")
+		}
+		// Overwrite, accept from this replica now.
+		cas = true
+	}
+
+	if now.Sub(timestamp.Time(entry.ReceivedAt)) > c.updateTimeout {
+		cas = true
+	}
+
+	if cas {
+		return c.casWrapper(ctx, cluster, replica, now)
+	}
+	return nil
+}
+
+// Returns an error if we don't want to accept the sample.
+func (c *tracker) checkKVStore(ctx context.Context, cluster, replica string, now time.Time) error {
+	return c.casWrapper(ctx, cluster, replica, now)
```

Comment:
Its a code smell to have a function which just call another one and does nothing else.  Can we remove this?  `s/casWrapper/checkKVStore/`